File: 1609-even-odd-tree/1609-even-odd-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isEvenOddTree(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        boolean evenLevel = true;
        while (!q.isEmpty()) {
            int n = q.size();
            int prev = evenLevel ? Integer.MIN_VALUE : Integer.MAX_VALUE;
            while (n-- > 0) {
                TreeNode temp = q.poll();
                if (temp.left != null) {
                    q.offer(temp.left);
                }
                if (temp.right != null) {
                    q.offer(temp.right);
                }
                if (evenLevel && (temp.val % 2 == 0 || temp.val <= prev)) {
                    return false;
                }
                if (!evenLevel && (temp.val % 2 != 0 || temp.val >= prev)) {
                    return false;
                }
                prev = temp.val;
            }
            evenLevel = !evenLevel;
        }
        return true;
    }
}

File: 1609-even-odd-tree/1609-even-odd-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        q = deque([root])
        even_level = True
        while q:
            n = len(q)
            prev = float('-inf') if even_level else float('inf')

            while n > 0:
                n -= 1
                temp = q.popleft()
                if temp.left:
                    q.append(temp.left)
                if temp.right:
                    q.append(temp.right)
                if even_level and (temp.val % 2 == 0 or temp.val <= prev):
                    return False
                if not even_level and (temp.val % 2 != 0 or temp.val >= prev):
                    return False
                prev = temp.val
            
            even_level = not even_level
        
        return True